src/core/traversal-cache.ts

```diff
@@ -0,0 +1,51 @@
+import { PartitionedCycles } from './cycles';
+
+type NodeId = string;
+
+type CacheMap<T> = Map<
+  NodeId,
+  {
+    item: T;
+
+    // the cycleNodeIds holds the nodes ids in a cycle
+    // i.e. for the cyclic graph "1->2->3->4->2", for nodeId=2 the cycleNodeIds will be "3,4"
+    // if nodeId exists in cycleNodeIds, don't use cached item version
+    cycleNodeIds?: Set<NodeId>;
+  }
+>;
+
+// This class adds cache when traversing over the graph
+// It will cache the item
+export class TraversalCache<T> {
+  private map: CacheMap<T> = new Map();
+
+  public set(nodeId: NodeId, item: T, partitionedCycles?: PartitionedCycles) {
+    if (!partitionedCycles) {
+      this.map.set(nodeId, { item });
+      return;
+    }
+    const { cyclesStartWithThisNode, cyclesWithThisNode } = partitionedCycles;
+    if (cyclesStartWithThisNode?.length > 0) {
+      // add explaination
+      const cycleNodeIds = new Set<NodeId>(...cyclesStartWithThisNode);
```

Comment:
Please don't spread here. We've run into stack overflows by doing that. Also, it's unnecessary as `cyclesStartWithThisNode` is an array.

```suggestion
      const cycleNodeIds = new Set<NodeId>(cyclesStartWithThisNode);
```
```suggestion
      const cycleNodeIds = new Set<NodeId>(...cyclesStartWithThisNode);
```